include/nbl/builtin/hlsl/bxdf/cook_torrance_base.hlsl

```diff
@@ -17,95 +17,303 @@ namespace hlsl
+
+        scalar_type rcpChoiceProb;
+        scalar_type z = u.z;
+        bool transmitted = math::partitionRandVariable(reflectionProb, z, rcpChoiceProb);
+
+        cache = anisocache_type::createForReflection(localV, H);
+
+        Refract<scalar_type> r = Refract<scalar_type>::create(localV, H);
+        cache.iso_cache.LdotH = hlsl::mix(cache.getVdotH(), r.getNdotT(rcpEta.value2[0]), transmitted);
+        ray_dir_info_type localL;
+        bxdf::ReflectRefract<scalar_type> rr;
+        rr.refract = r;
+        localL = localL.reflectRefract(rr, transmitted, rcpEta.value[0]);
+
+        // fail if samples have invalid paths
+        if ((!transmitted && hlsl::sign(localL.getDirection().z) != hlsl::sign(localV.z)) || (transmitted && hlsl::sign(localL.getDirection().z) == hlsl::sign(localV.z)))
         {
-            measure_transform.pdf *= ndf.template correlated<Query, sample_type, anisotropic_interaction_type>(query, _sample, interaction);
+            localL.direction = vector3_type(0,0,0); // should check if sample direction is invalid
         }
-        return measure_transform;
+
+        return sample_type::createFromTangentSpace(localL, interaction.getFromTangentSpace());
+    }
+
+    scalar_type pdf(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
+    {
+        using quant_query_type = typename N::quant_query_type;
+        using dg1_query_type = typename N::dg1_query_type;
+
+        fresnel::OrientedEtas<monochrome_type> orientedEta = fresnel.orientedEta;
+        quant_query_type qq = ndf.template createQuantQuery<isocache_type>(cache, orientedEta.value[0]);
+        dg1_query_type dq = ndf.template createDG1Query<isotropic_interaction_type, isocache_type>(interaction, cache);
+        quant_type DG1 = ndf.template DG1<sample_type, isotropic_interaction_type>(dq, qq, _sample, interaction);
+
+        const scalar_type reflectance = fresnel(hlsl::abs(cache.getVdotH()))[0];
+        return hlsl::mix(reflectance, scalar_type(1.0) - reflectance, cache.isTransmission()) * DG1.projectedLightMeasure;
+    }
+    scalar_type pdf(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(anisocache_type) cache)
+    {
+        using quant_query_type = typename N::quant_query_type;
+        using dg1_query_type = typename N::dg1_query_type;
+
+        fresnel::OrientedEtas<monochrome_type> orientedEta = fresnel.orientedEta;
+        quant_query_type qq = ndf.template createQuantQuery<anisocache_type>(cache, orientedEta.value[0]);
+        dg1_query_type dq = ndf.template createDG1Query<anisotropic_interaction_type, anisocache_type>(interaction, cache);
+        quant_type DG1 = ndf.template DG1<sample_type, anisotropic_interaction_type>(dq, qq, _sample, interaction);
```

Comment:
you need to reciprocate the fresnel's oriented Eta depending on sign of `V` here and in `eval`, `generate` and all the other functions, don't count on the user doing that for you if you define a BSDF.

Unless we change the definitions of all BxDFs so they're only defined for `VdotN>0` and keep that convention throughout all the code.